From: Konstantin Khlebnikov <khlebnikov@openvz.org>
Date: Tue Sep 24 15:27:42 2013 -0700
Subject: audit: fix endless wait in audit_log_start()
Git-commit: 8ac1c8d5deba65513b6a82c35e89e73996c8e0d6
Patch-mainline: v3.12-rc3
References: bnc#908393
Signed-off-by: Tony Jones <tonyj@suse.de>

    audit: fix endless wait in audit_log_start()
    
    After commit 829199197a43 ("kernel/audit.c: avoid negative sleep
    durations") audit emitters will block forever if userspace daemon cannot
    handle backlog.
    
    After the timeout the waiting loop turns into busy loop and runs until
    daemon dies or returns back to work.  This is a minimal patch for that
    bug.
    
    Signed-off-by: Konstantin Khlebnikov <khlebnikov@openvz.org>
    Cc: Luiz Capitulino <lcapitulino@redhat.com>
    Cc: Richard Guy Briggs <rgb@redhat.com>
    Cc: Eric Paris <eparis@redhat.com>
    Cc: Chuck Anderson <chuck.anderson@oracle.com>
    Cc: Dan Duval <dan.duval@oracle.com>
    Cc: Dave Kleikamp <dave.kleikamp@oracle.com>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

diff --git a/kernel/audit.c b/kernel/audit.c
index 91e53d0..7b0e23a 100644
--- a/kernel/audit.c
+++ b/kernel/audit.c
@@ -1117,9 +1117,10 @@ struct audit_buffer *audit_log_start(struct audit_context *ctx, gfp_t gfp_mask,
 
 			sleep_time = timeout_start + audit_backlog_wait_time -
 					jiffies;
-			if ((long)sleep_time > 0)
+			if ((long)sleep_time > 0) {
 				wait_for_auditd(sleep_time);
-			continue;
+				continue;
+			}
 		}
 		if (audit_rate_check() && printk_ratelimit())
 			printk(KERN_WARNING
